### PR TITLE
Pipeline pods volumes

### PR DESCRIPTION
**Pull request by michalzelechowski-getindata**
Keeping only persistent volume claims for now. Support of RWO policy (works well when pods are allocated from the same node pool, otherwise wait for start). Storage size is configurable.

### CODE REVIEW

**Review by em-pe — APPROVED** [review]
🥇

File: kedro_airflow_k8s/airflow_dag_template.j2 [hunk 1]
```diff
@@ -33,6 +37,34 @@ with DAG(
         python_callable=start_mlflow_run,
     )
 
+    def create_pipeline_storage(ti, **kwargs):
+        with client.ApiClient(config.load_incluster_config()) as api_client:
+            v1 = client.CoreV1Api(api_client)
+
+            pvc = k8s.V1PersistentVolumeClaim(metadata=k8s.V1ObjectMeta(name=pvc_name, namespace='{{ config.namespace }}'),
+                    spec=k8s.V1PersistentVolumeClaimSpec(access_modes=['ReadWriteOnce'],
```
Comment (szczeles) [inline kedro_airflow_k8s/airflow_dag_template.j2:45]: Consider pushing access modes to config - some setups may have for example NFS that allows ReadWriteMany

File: kedro_airflow_k8s/airflow_dag_template.j2 [hunk 2]
```diff
@@ -33,6 +37,34 @@ with DAG(
...
+
+    def delete_pipeline_storage(ti, **kwargs):
+        with client.ApiClient(config.load_incluster_config()) as api_client:
+            v1 = client.CoreV1Api(api_client)
+            v1.delete_namespaced_persistent_volume_claim(name=pvc_name, namespace='{{ config.namespace }}')
+
+    delete_pipeline_storage = PythonOperator(
+        task_id='delete_pipeline_storage',
+        python_callable=delete_pipeline_storage,
+        trigger_rule='all_done'
```
Comment (szczeles) [inline kedro_airflow_k8s/airflow_dag_template.j2:63]: That is simply amazing! :rocket:

File: kedro_airflow_k8s/airflow_dag_template.j2 [hunk 3]
```diff
@@ -33,6 +37,34 @@ with DAG(
...
+    )
+
+    volume = k8s.V1Volume(name='storage',persistent_volume_claim=k8s.V1PersistentVolumeClaimVolumeSource(claim_name=pvc_name))
+    volume_mount = k8s.V1VolumeMount(name='storage', mount_path='/data', sub_path=None, read_only=False)
```
Comment (szczeles) [inline kedro_airflow_k8s/airflow_dag_template.j2:67]: Consider mounting into /home/kedro/data -> this is where Kedro expects the data in the standard image

File: kedro_airflow_k8s/airflow_dag_template.j2 [hunk 4]
```diff
@@ -20,6 +22,8 @@ with DAG(
     params={"example_key": "example_value"},
 ) as dag:
 
+    pvc_name = '{{ project_name | safe | slugify }}.{{ execution_marker }}'
```
Comment (szczeles) [inline kedro_airflow_k8s/airflow_dag_template.j2:25]: Consider using `{{ ts_nodash }}` (https://airflow.apache.org/docs/apache-airflow/stable/macros-ref.html) -> it needs to be set on **execution** (to be random for different runs rather during pipeline generation (that may be done only once)

File: kedro_airflow_k8s/cli.py [hunk 5]
```diff
@@ -55,6 +56,13 @@ def generate(ctx, target_path="dags/"):
             dependencies[parent].append(node)
             nodes_with_no_deps = nodes_with_no_deps - set([node.name])
 
+    all_parent_nodes = set()
+    for _, parent_nodes in pipeline.node_dependencies.items():
+        all_parent_nodes = all_parent_nodes.union(
+            set(parent.name for parent in parent_nodes)
+        )
+    bottom_nodes = set(node.name for node in pipeline.nodes) - all_parent_nodes
```
Comment (szczeles) [inline kedro_airflow_k8s/cli.py:64]: Fantastic!

**Comment by szczeles** [comment]
@michalzelechowski-getindata It looks really good! I left a small minor comments

File: kedro_airflow_k8s/airflow_dag_template.j2 (hunk 2 above)
Comment (michalzelechowski-getindata) [reply kedro_airflow_k8s/airflow_dag_template.j2:63]: :bow:

File: kedro_airflow_k8s/airflow_dag_template.j2 (hunk 3 above)
Comment (michalzelechowski-getindata) [reply kedro_airflow_k8s/airflow_dag_template.j2]: Thanks for pointing that out.

File: kedro_airflow_k8s/cli.py (hunk 5 above)
Comment (michalzelechowski-getindata) [reply kedro_airflow_k8s/cli.py:64]: :bow:

File: kedro_airflow_k8s/airflow_dag_template.j2 (hunk 1 above)
Comment (michalzelechowski-getindata) [reply kedro_airflow_k8s/airflow_dag_template.j2]: Good point, I am changing this.

File: kedro_airflow_k8s/airflow_dag_template.j2 (hunk 4 above)
Comment (michalzelechowski-getindata) [reply kedro_airflow_k8s/airflow_dag_template.j2:25]: Good catch, and that is what I was looking for :) Is it just sufficient to put it in place and Airflow will replace {{ ts_nodash }} during the execution - that's what you meant, right?

File: README.md [hunk 6]
```diff
@@ -12,22 +12,28 @@
...
+```
+Usage: kedro airflow-k8s [OPTIONS] COMMAND [ARGS]...
+ 
+Options:
+  -e, --env TEXT  Environment to use.
+  -h, --help      Show this message and exit.
 
-## Configuration file
+Commands:
+  generate  Create an Airflow DAG for a project
```
Comment (em-pe) [inline README.md:37]: Maybe change `generate` to `compile` to match the naming convention with kedro-kubeflow plugin? @szczeles - what do you think?

**Review by em-pe — APPROVED** [review]

File: kedro_airflow_k8s/airflow_dag_template.j2 (hunk 4 above)
Comment (szczeles) [reply kedro_airflow_k8s/airflow_dag_template.j2]: Indeed. If I'm not mistaken, all the parameters pushed to PythonOperator are templated, so it should work OOTB

File: README.md (hunk 6 above)
Comment (szczeles) [reply README.md:37]: I like the idea!

Comment (michalzelechowski-getindata) [reply README.md:37]: Ok, done.

**Comment by em-pe** [comment]
@michalzelechowski-getindata - when merging please 'squash & merge'

**Pull request merged**